drivers/scheduler/anthos/anthos.go

```diff
@@ -0,0 +1,576 @@
+	k8sOps, err := core.NewInstanceFromConfigFile(path.Join(anth.confPath, kubeConfig))
+	if err != nil {
+		return err
+	}
+	adminNodeList, err := k8sOps.GetNodes()
+	if err != nil {
+		return err
+	}
+	adminIndex := startAdminNodeIndex
+	for _, adminNode := range adminNodeList.Items {
+		if adminIndex > lastAdminNodeIndex {
+			break
+		}
+		anth.instances[adminIndex].HostName = adminNode.Name
+		anth.instances[adminIndex].PublicIpAddress = adminNode.Status.Addresses[0].Address
+		anth.instances[adminIndex].PrivateIpAddress = adminNode.Status.Addresses[0].Address
+		adminIndex += 1
+	}
+	return nil
+}
+
+// saveInstance save the instances.json after upgrade
+func (anth *anthos) saveInstance() error {
+	log.Debug("Saving instances.json")
+	b, err := json.MarshalIndent(anth.instances, "", "  ")
+	if err != nil {
+		return fmt.Errorf("cannot marshall Instances, %v", err)
+	}
+	if os.WriteFile(path.Join(anth.instPath, jsonInstances), b, 0644); err != nil {
+		return fmt.Errorf("couldn't write to %s/%s: %v", anth.instPath, jsonInstances, err)
+	}
+	return nil
+}
+
+// verifyUpgradeVersion validates that correct version is provided for upgrade or not
+func (anth *anthos) verifyUpgradeVersion(upgradeVersion string) error {
+	log.Infof("Checking the upgrade from version: [%s] to version: [%s]",
+		anth.version, upgradeVersion)
+	var vReg = regexp.MustCompile(`(^\w).(\w+).(\w+)`)
+	parseV1 := strings.Split(anth.version, "-")
+	parseV2 := strings.Split(upgradeVersion, "-")
+	v1 := strings.TrimSpace(parseV1[0])
+	v2 := strings.TrimSpace(parseV2[0])
+	version1, err := version.NewVersion(v1)
+	if err != nil {
+		return err
+	}
+	version2, err := version.NewVersion(v2)
+	if err != nil {
+		return err
+	}
+	if version1.GreaterThanOrEqual(version2) {
+		return fmt.Errorf("incorrect upgrade version:%s is provided."+
+			"Upgrade version should be higher", upgradeVersion)
+	}
+	toVersion := vReg.FindAllStringSubmatch(v1, -1)
+	fromVersion := vReg.FindAllStringSubmatch(v2, -1)
+	val1, err := strconv.Atoi(toVersion[0][2])
+	if err != nil {
+		return err
+	}
+	val2, err := strconv.Atoi(fromVersion[0][2])
+	if err != nil {
+		log.Errorf("Failed to parse version: %s", fromVersion)
+		return err
+	}
+	if (len(toVersion) > 0 && len(fromVersion) > 0) &&
+		(toVersion[0][1] != fromVersion[0][1] || (val2-val1) > 1) {
+		return fmt.Errorf("incorrect upgrade version:%s is provided."+
+			"One major version upgrade support at a time", upgradeVersion)
+	}
+	log.Debugf("Successfully verified the version:[%]", upgradeVersion)
+	return nil
+}
+
+// updateGkeadmUtil update gkeadm version to given version
+func (anth *anthos) updateGkeadmUtil(version string) error {
+	log.Infof("Updating gkeadm to version: [%s]", version)
+	src := fmt.Sprintf("gs://gke-on-prem-release/gkeadm/%s/linux/gkeadm", version)
+	if out, err := exec.Command(gsUtilCmd, "cp", src, anth.confPath).CombinedOutput(); err != nil {
+		log.Errorf("Copying gkeadm from src: [%s] failed with error: %v", src, err)
```

Comment:
you can combine this error statement and return it, so that we can avoid double logging of the error

Reply:
Done